language-support/ts/daml-ledger/index.ts

```diff
@@ -608,44 +628,147 @@ class Ledger {
   }
 
   /**
    * Retrieve a consolidated stream of events for a given template and contract key.
    *
-   * Same as [[streamQuery]], but instead of a query, match contracts by contract key.
+   * The accumulated state is either the current active contract for the given
+   * key, or null if there is no active contract for the given key.
    *
    * @typeparam T The contract template type.
    * @typeparam K The contract key type.
    * @typeparam I The contract id type.
+   *
    */
   streamFetchByKey<T extends object, K, I extends string>(
     template: Template<T, K, I>,
     key: K,
   ): Stream<T, K, I, CreateEvent<T, K, I> | null> {
-    let lastContractId: ContractId<T> | null = null;
-    const request = [{templateId: template.templateId, key}];
-    const reconnectRequest = (): object[] => [{...request[0], 'contractIdAtOffset': lastContractId}]
-    const change = (contract: CreateEvent<T, K, I> | null, events: readonly Event<T, K, I>[]): CreateEvent<T, K, I> | null => {
-      // NOTE(MH, #4564): We're very lenient here. We should not see a create
-      // event when `contract` is currently not null. We should also only see
-      // archive events when `contract` is currently not null and the contract
-      // ids match. However, the JSON API does not provied these guarantees yet
-      // but we're working on them.
+    return this.streamFetchByKeyCommon(template, [key], "streamFetchByKey", (cs) => cs[0], (c) => [c]);
+  }
+
+  /**
+   * @internal
+   *
+   * Returns the same API as [[streamSubmit]] but does not, in fact, establish
+   * any socket connection. Instead, this is a stream that always has the given
+   * value as its accumulated state.
+   */
+  constantStream<T extends object, K, I extends string, V>(
+    value: V,
+  ): Stream<T, K, I, V> {
+    function on(t: 'live', l: (v: V) => void): void;
+    function on(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function on(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function on(type: any, listener: any): any {
+        if (type === 'live') {
+          listener(value);
+        }
+        if (type === 'change') {
+          listener(value, []);
+        }
+    }
+    function off(t: 'live', l: (v: V) => void): void;
+    function off(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function off(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+    function off(_t: any, _l: any): any {}
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { on, off, close: (): void => {} };
+  }
+
+  /**
+   * @internal
+   *
+   */
+  streamFetchByKeyCommon<T extends object, K, I extends string, S>(
+    template: Template<T, K, I>,
+    keys: K[], // non-zero length
```

Comment:
Worth adding a `console.assert` for?

Reply:
I don't think so. This is a private method se we're in control of all of its two call sites.

I guess it would be even more private if I added the `private` keyword, though.

Reply:
Might be a difference in how we use assertions. Personally, I tend to use assertions for internal invariants, i.e., things that should always hold but could be broken accidentally by a refactoring so it doesn’t really matter if it’s private or not.

Reply:
I don’t feel strongly about this so happy to let you make the decision.